zebra-rpc/src/server/rpc_call_compatibility.rs

```diff
@@ -3,116 +3,66 @@
-        if let Some(Output::Failure(Failure { ref mut error, .. })) = output {
-            if matches!(error.code, jsonrpc_core::ErrorCode::InvalidParams) {
-                let original_code = error.code.clone();
-
-                error.code = server::error::LegacyCode::Misc.into();
-                tracing::debug!("Replacing RPC error: {original_code:?} with {error}");
-            }
-        }
+    /// Create a new `FixRpcResponseMiddleware` with the given `service`.
+    pub fn new(service: RpcService) -> Self {
+        Self { service }
     }
+}
 
-    /// Obtain a description string for a received request.
-    ///
-    /// Prints out only the method name and the received parameters.
-    fn call_description(call: &Call) -> String {
-        const MAX_PARAMS_LOG_LENGTH: usize = 100;
+impl<'a> RpcServiceT<'a> for FixRpcResponseMiddleware {
+    type Future = ResponseFuture<futures::future::BoxFuture<'a, jsonrpsee::MethodResponse>>;
 
-        match call {
-            Call::MethodCall(MethodCall { method, params, .. }) => {
-                let mut params = format!("{params:?}");
-                if params.len() >= MAX_PARAMS_LOG_LENGTH {
-                    params.truncate(MAX_PARAMS_LOG_LENGTH);
-                    params.push_str("...");
-                }
+    fn call(&self, request: jsonrpsee::types::Request<'a>) -> Self::Future {
+        let service = self.service.clone();
+        ResponseFuture::future(Box::pin(async move {
+            let response = service.call(request).await;
+            if response.is_error() {
+                let original_error_code = response
+                    .as_error_code()
+                    .expect("response should have an error code");
+                if original_error_code == jsonrpsee_types::ErrorCode::InvalidParams.code() {
+                    let new_error_code = crate::server::error::LegacyCode::Misc.into();
+                    tracing::debug!(
+                        "Replacing RPC error: {original_error_code} with {new_error_code}"
+                    );
+                    let json: serde_json::Value =
+                        serde_json::from_str(response.into_parts().0.as_str())
+                            .expect("response string should be valid json");
+                    let id = json["id"]
+                        .as_str()
+                        .expect("response json should have an id")
+                        .to_string();
 
-                format!(r#"method = {method:?}, params = {params}"#)
-            }
-            Call::Notification(Notification { method, params, .. }) => {
-                let mut params = format!("{params:?}");
-                if params.len() >= MAX_PARAMS_LOG_LENGTH {
-                    params.truncate(MAX_PARAMS_LOG_LENGTH);
-                    params.push_str("...");
+                    return MethodResponse::error(
+                        jsonrpsee_types::Id::Str(id.into()),
```

Comment:
It may still be possible to only replace the error code, which would be simpler.

Reply:
I agree. I opened https://github.com/ZcashFoundation/zebra/issues/9097 for it.